libraries/c_sdk/standard/ble/src/iot_ble_mqtt_transport.c

```diff
@@ -0,0 +1,1081 @@
+
+
+bool IotBleMqttTransportInit( const NetworkContext_t * context )
+{
+    bool status = true;
+
+    streamBuffer = xStreamBufferCreateStatic( context->bufSize, 1, context->buf, &xStreamBufferStruct );
+
+    if ( streamBuffer == NULL )
+    {
+        status = false;
+    }
+
+    return status;
+}
+
+
+void IotBleMqttTransportCleanup( void )
+{
+    vStreamBufferDelete( streamBuffer );
+}
+
+
+static uint16_t getIntFromTwoBytes( const uint8_t * buf )
+{
+    uint16_t result = 0;
+
+    result = buf[ 0 ];
+    result <<= 8;
+    result += buf[ 1 ];
+    return result;
+}
+
+
+static MQTTQoS_t convertIntToQos( const uint8_t incomingQos )
+{
+    MQTTQoS_t qosValue = MQTTQoS0;
+    
+    switch( incomingQos )
+    {
+        case 0U:
+            qosValue = MQTTQoS0;
+            break;
+            
+        case 1U:
+            qosValue = MQTTQoS1;
+            break;
+            
+        default:
+            LogError( ( "QoS 2 is not supported by MQTT over BLE. Defaulting to Qos 1." ) );
+            qosValue = MQTTQoS1;
+            break;
+    }
+    
+    return qosValue;
+}
+
+
+static MQTTStatus_t parseConnect( MQTTConnectInfo_t * connectConfig,
+                                  const uint8_t * buf,
+                                  size_t * pSize )
+{
+    bool willFlag = false;
+    bool willRetain = false;
+    bool passwordFlag = false;
+    bool usernameFlag = false;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTQoS_t willQos = 0;
+    uint8_t connectionFlags = 0;
+    size_t bufferIndex = 0;
+    size_t payloadIndex = 0;
+
+
+    if( ( buf[ 0 ] & 0xf0U ) != MQTT_PACKET_TYPE_CONNECT )
+    {
+        status = MQTTBadParameter;
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Skips the remaining length, which is not needed in connect packet.  Finds first 'M' (77) in 'MQTT'. */
+        while( buf[ bufferIndex ] != 77U )
+        {
+            bufferIndex++;
+        }
+
+        /* Skips 'MQTT'. */
+        bufferIndex += 4U;
+
+        /* The service level of the packet. Must be 4. */
+        if( ( buf[ bufferIndex ] != 4U ) )
+        {
+            LogError( ( "The service level of a connect packet must be 4, see [MQTT-3.1.2-2]." ) );
+            status = MQTTBadParameter;
+        }
+
+        /* Increment bufferIndex to be the start of the payload. */
+        bufferIndex++;
+        /* Second byte after payload is connection flags. */
+        connectionFlags = buf[ bufferIndex ];
+
+        /* The LSB is reserved and must be 0. */
+        if( ( connectionFlags & 0x01U ) != 0U )
+        {
+            LogError( ( "LSB of Connect Flags byte must be 0, see [MQTT-3.1.2-3]." ) );
+            status = MQTTBadParameter;
+        }
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Get flag data from the flag byte. */
+        connectConfig->cleanSession = ( ( ( connectionFlags & 0x02U ) >> 1) == 1U );
+        willFlag = ( ( ( connectionFlags & 0x04U ) >> 2 ) == 1U) ;
```

Comment:
This and some of other checks could be replaced by defining masks. Define a `WILL_FLAG_MASK` to `0x04`. It will avoid magic numbers in code as well.